hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -206,73 +286,116 @@ public void process() throws Exception {
+                    break;
+                }
+            } else if (state == State.BLOCKED) {
+                // it is already blocked, so we are done.
+                break;
+            } else if (state == State.RESCHEDULE) {
+                // rescheduling is requested, so lets do that. Once put to RESCHEDULE,
+                // only the IO thread will change the state, so we can safely call a set.
+                scheduled.set(State.SCHEDULED);
+                // this will cause the pipeline to be rescheduled.
+                owner().addTaskAndWakeup(this);
+                break;
+            } else {
+                throw new IllegalStateException("unexpected state:" + state);
+            }
         }
+    }
 
-        // So there are frames, but we just unscheduled ourselves. If we don't try to reschedule, then these
-        // Frames are at risk not to be send.
-        if (!scheduled.compareAndSet(false, true)) {
-            //someone else managed to schedule this OutboundHandler, so we are done.
-            return;
+    private void postProcessDirty() throws IOException {
+        // pipeline is dirty, so register for an OP_WRITE to write more data.
+        registerOp(OP_WRITE);
+
+        if (writeThroughEnabled && !(Thread.currentThread() instanceof NioThread)) {
+            // there was a write through. Changing the interested set of the selection key
+            // after the IO thread did a select, will not lead to the selector waking up. So
+            // if we don't wake up the selector explicitly, only after the selector.select(timeout)
+            // has expired the selectionKey will be seen. For more info see:
+            // https://stackoverflow.com/questions/11523471/java-selectionkey-interestopsint-not-thread-safe
+            owner.getSelector().wakeup();
+            concurrencyDetection.onDetected();
         }
+    }
 
-        // We managed to reschedule. So lets add ourselves to the owner so we are processed again.
-        // We don't need to call wakeup because the current thread is the IO-thread and the selectionQueue will be processed
-        // till it is empty. So it will also pick up tasks that are added while it is processing the selectionQueue.
+    private void postProcessClean() throws IOException {
+        // There is nothing left to be done; so lets unschedule this pipeline
+        // since everything is written, we are not interested anymore in write-events, so lets unsubscribe
+        unregisterOp(OP_WRITE);
 
-        // owner can't be null because this method is made by the owning io thread.
-        owner().addTask(this);
+        for (; ; ) {
+            State state = scheduled.get();
+            if (state == State.RESCHEDULE) {
+                if (!scheduled.compareAndSet(State.RESCHEDULE, State.SCHEDULED)) {
+                    continue;
+                }
+                owner().addTaskAndWakeup(this);
```

Comment:
Same as the first comment - since this method can be called concurrently, I guess the first thread may reschedule the task after which the second thread may set the state to unscheduled. Can this be a problem?

Reply:
```
private void postProcessClean() throws IOException {
        // There is nothing left to be done; so lets unschedule this pipeline
        // since everything is written, we are not interested anymore in write-events, so lets unsubscribe
        unregisterOp(OP_WRITE);

        for (; ; ) {
            State state = scheduled.get();
            if (state == State.RESCHEDULE) {
                // the pipeline needs to be rescheduled. The current thread is still owner of the pipeline,
                // so lets remove the reschedule flag and return it to schedule and lets reprocess the pipeline.
                scheduled.set(State.SCHEDULED);
                owner().addTaskAndWakeup(this);
                return;
            }

            if (!scheduled.compareAndSet(state, State.UNSCHEDULED)) {
                // we didn't manage to set it to unscheduled, lets retry the loop and see what needs to be done
                continue;
            }

            // ---------- from this point it is potentially concurrent

            // we manage to unschedule the pipeline. From this point on we have released ownership of the pipeline
            // and another thread could call the process method. 
            if (writeQueue.isEmpty() && priorityWriteQueue.isEmpty()) {
                //pipeline is clean, we are done.
                return;
            }

            // there is stuff to write we are going to reclaim ownership of the pipeline to prevent
            // we are going to end up with scheduled pipeline that isn't going to be processed.
            // If we can't reclaim ownership, then it is the concern of the other thread deal with the pipeline.
            if (scheduled.compareAndSet(State.UNSCHEDULED, State.SCHEDULED)) {
                if (Thread.currentThread().getClass() == NioThread.class) {
                    owner().addTask(this);
                } else {
                    owner().addTaskAndWakeup(this);
                }
            }

            return;
        }
```

If the first thread reschedules, the first thread is going to be owner. If the second thread scheduled, he becomes owner. You can't have two owners of the pipeline.